misc/testerina/modules/testerina-runtime/src/main/java/org/ballerinalang/test/runtime/entity/CoverageReport.java

```diff
@@ -76,60 +80,94 @@ public CoverageReport(Path sourceJarPath, Path targetDirPath, String orgName, St
      * @throws IOException when file operations are failed
      */
     public void generateReport() throws IOException {
-        try {
-            CodeCoverageUtils.unzipCompiledSource(sourceJarPath, projectDir, orgName, moduleName, version);
-        } catch (NoSuchFileException e) {
+
+        // Obtain a path list of all the .jar files generated
+        List<Path> pathList;
+        try (Stream<Path> walk = Files.walk(this.jarCache, 5)) {
+            pathList = walk.map(path -> path).filter(f -> f.toString().endsWith(version + ".jar")).collect(
+                    Collectors.toList());
+        } catch (IOException e) {
             return;
         }
 
-        execFileLoader.load(executionDataFile.toFile());
+        if (!pathList.isEmpty()) {
+            // For each jar file found, we unzip it for this particular module
+            for (Path moduleJarPath : pathList) {
+                try {
+                    // Creates coverage folder with each class per module
+                    CodeCoverageUtils.unzipCompiledSource(moduleJarPath, projectDir, orgName, moduleName, version);
+                } catch (NoSuchFileException e) {
+                    return;
+                }
+            }
+
+            execFileLoader.load(executionDataFile.toFile());
+            final IBundleCoverage bundleCoverage = analyzeStructure();
+            createReport(bundleCoverage);
+        } else {
+            String msg = "Unable to generate code coverage for the module " + moduleName + ". Jar files dont exist.";
+            throw new NoSuchFileException(msg);
+        }
 
-        final IBundleCoverage bundleCoverage = analyzeStructure();
-        createReport(bundleCoverage);
     }
 
     private IBundleCoverage analyzeStructure() throws IOException {
         final CoverageBuilder coverageBuilder = new CoverageBuilder();
-        final Analyzer analyzer = new Analyzer(
-                execFileLoader.getExecutionDataStore(), coverageBuilder);
+        final Analyzer analyzer = new Analyzer(execFileLoader.getExecutionDataStore(), coverageBuilder);
         analyzer.analyzeAll(classesDirectory.toFile());
         return coverageBuilder.getBundle(title);
     }
 
     private void createReport(final IBundleCoverage bundleCoverage) {
-        boolean containsSourceFiles;
+        boolean containsSourceFiles = true;
 
         for (IPackageCoverage packageCoverage : bundleCoverage.getPackages()) {
+            // I havent tested the behaviour of single files
```

Comment:
Is this a TODO?

Reply:
Code coverage is actually not supported for single files. I checked. I will remove this TODO comment